### PR TITLE
Feature: Adds communication between Panels

### LINKED ISSUE #19432: POC: Investigate the best way to communicate between Panels in a Dashboard
We need a typesafe way to communicate events between Panels on a Dashboard. This could probably use some of the amazing groundwork being done here #19418.

**Open questions investigations**
- Using React Context?
- Using Redux?
- Using props?

**Some requirements:**
1. Must be able to communicate with all Panel Plugins, React and Angular
2. Should not expose consumers to difficult api or usage of the same, as `unsubscribing`.
3. A Panel should be able to receive and send events.

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
This is parallel PR to #19468 and is now independent of the amazing work done here: #19418

**Which issue(s) this PR fixes**:

Closes #19432

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/core/hooks/useEventing.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+export interface Eventing {
+  subscribeToEvents: (options: { tap: (event: Event<any>) => void; filter?: (event: Event<any>) => boolean }) => void;
+  publishEvent: (event: Event<any>) => void;
+}
+
+export interface Event<T extends {} = {}> {
+  name?: string;
+  origin?: string;
+  payload?: T;
+}
```
Comment (hugohaggmark) [inline public/app/core/hooks/useEventing.ts]: Temporary use until `event_types` branch is merged.

File: public/app/features/dashboard/dashgrid/DashboardPanel.tsx [hunk 2]
```diff
@@ -90,7 +88,20 @@ export class DashboardPanel extends PureComponent<Props, State> {
...
+        if (event.origin === panel.id.toString()) {
+          return false;
+        }
+
+        return event.name === 'mouse-move-event';
+      },
+      tap: event => {
+        console.log(`New event arrived from ${event.origin} at ${panel.id}`, event);
+      },
+    });
```
Comment (hugohaggmark) [inline public/app/features/dashboard/dashgrid/DashboardPanel.tsx]: For test purposes to see that this actually works :)

File: public/app/features/dashboard/dashgrid/DashboardPanel.tsx [hunk 3]
```diff
@@ -190,7 +201,12 @@ export class DashboardPanel extends PureComponent<Props, State> {
     });
 
     return (
-      <div className={editorContainerClasses}>
+      <div
+        className={editorContainerClasses}
+        onMouseMove={event =>
+          this.props.publishEvent({ name: 'mouse-move-event', origin: this.props.panel.id.toString(), payload: event })
+        }
```
Comment (hugohaggmark) [inline public/app/features/dashboard/dashgrid/DashboardPanel.tsx]: For test purposes to see that this actually works :)

**Review by ryantxu — COMMENTED** [review]
This is looking interesting!

what are your thoughts on how this gets exposed to PanelChrome/PanelComponent(plugin.panel)?  Are you proposing that we add:
```
  publishEvent={publishEvent}
  subscribeToEvents={subscribeToEvents}
```
to everything?  or that a plugin could/should directly call:
```
<EventsConsumer>
          {({ publishEvent, subscribeToEvents }) => (
          ...
</EventsConsumer>
```

Assuming the <EventConsumer is lightweight, having properties seems easiest.  But note that most plugins will not use the events.

**Comment by hugohaggmark** [comment]
> This is looking interesting!
> 
> what are your thoughts on how this gets exposed to PanelChrome/PanelComponent(plugin.panel)? Are you proposing that we add:
> 
> ```
>   publishEvent={publishEvent}
>   subscribeToEvents={subscribeToEvents}
> ```
> 
> to everything? or that a plugin could/should directly call:
> 
> ```
> <EventsConsumer>
>           {({ publishEvent, subscribeToEvents }) => (
>           ...
> </EventsConsumer>
> ```
> 
> Assuming the <EventConsumer is lightweight, having properties seems easiest. But note that most plugins will not use the events.

Good question, as this is in the React Context any component can pull it in so using the `EventConsumer` will make that easier for any component wanting to take advantage of this.

**Comment by ryantxu** [comment]
Maybe `<EventConsumer` should be something else since it is is also a producer.  Hugo suggested `<Events` or `<WithEvents`

If we want plugins to use `<WithEvents` directly, it should be moved to `@grafana/ui` -- otherwise publish & subscribe should be added to plugin properties.

**Comment by hugohaggmark** [comment]
Will close this in favor of #20720

**Pull request closed without merging**